cli/flox-activations/src/cli/activate.rs

```diff
@@ -0,0 +1,219 @@
+                ShellWithPath::Zsh(_) => format!("export {key}={value};",),
+            })
+            .join("\n");
+
+        let script = formatdoc! {"
+            {exports_rendered}
+            {output}
+        ",
+        output = String::from_utf8_lossy(&output.stdout),
+        };
+
+        print!("{script}");
+
+        Ok(())
+    }
+
+    fn assemble_command_for_activate_script(data: ActivateCtx) -> Command {
+        let mut exports = HashMap::from([
+            (FLOX_ACTIVE_ENVIRONMENTS_VAR, data.flox_active_environments),
+            (FLOX_ENV_LOG_DIR_VAR, data.flox_env_log_dir),
+            ("FLOX_PROMPT_COLOR_1", data.prompt_color_1),
+            ("FLOX_PROMPT_COLOR_2", data.prompt_color_2),
+            // Set `FLOX_PROMPT_ENVIRONMENTS` to the constructed prompt string,
+            // which may be ""
+            (FLOX_PROMPT_ENVIRONMENTS_VAR, data.flox_prompt_environments),
+            ("_FLOX_SET_PROMPT", data.set_prompt.to_string()),
+            ("_FLOX_ACTIVATE_STORE_PATH", data.flox_activate_store_path),
+            (
+                // TODO: we should probably figure out a more consistent way to
+                // pass this since it's also passed for `flox build`
+                FLOX_RUNTIME_DIR_VAR,
+                data.flox_runtime_dir,
+            ),
+            ("_FLOX_ENV_CUDA_DETECTION", data.flox_env_cuda_detection),
+            (
+                FLOX_ACTIVATE_START_SERVICES_VAR,
+                data.flox_activate_start_services.to_string(),
+            ),
+            (FLOX_SERVICES_SOCKET_VAR, data.flox_services_socket),
+        ]);
+
+        if let Some(services_to_start) = data.flox_services_to_start {
+            exports.insert(FLOX_SERVICES_TO_START_VAR, services_to_start);
+        }
+
+        exports.extend(default_nix_env_vars());
+
+        let activate_path = data.interpreter_path.join("activate");
+        let mut command = Command::new(activate_path);
+        command.envs(exports);
+
+        command.arg("--env").arg(&data.env);
+        command
+            .arg("--env-project")
+            .arg(data.env_project.to_string_lossy().to_string());
+        command
+            .arg("--env-cache")
+            .arg(data.env_cache.to_string_lossy().to_string());
+        command.arg("--env-description").arg(data.env_description);
+
+        // Pass down the activation mode
+        command.arg("--mode").arg(data.mode);
+
+        command
+            .arg("--watchdog")
+            .arg(data.watchdog_bin.to_string_lossy().to_string());
+
+        command.arg("--shell").arg(data.shell.exe_path());
+
+        command
+    }
+
+    /// Quote run args so that words don't get split,
+    /// but don't escape all characters.
+    ///
+    /// To do this we escape '"' and '`',
+    /// but we don't escape anything else.
+    /// We want '$' for example to be expanded by the shell.
+    fn quote_run_args(run_args: &[String]) -> String {
```

Comment:
Nit: this probably belongs in the `shell_gen` module or `flox-core::shell` so it can be used in other contexts.

Reply:
Agreed but I'll merge for now